### ISSUE TITLE
Consider Updating [profile.release] to shrink binary size

### ISSUE THREAD

**Issue by smklein** [oxidecomputer/cli-old#155]
I ran the build with a handful of different configuration options, to try to cut down the current release-mode binary output size.

Below are my results, building with a handful of different options - the changes in output binary size are significant:

## Debug build
Just `cargo build`:
The `oxide` binary ends up being 246.1 MB

## Release build
`cargo build --release`, using: https://github.com/oxidecomputer/cli/blob/2fdc63e4a7367dc4633fed808b0e6baad34e0e77/Cargo.toml#L82-L83
216.3 MB

## Enabling LTO
`cargo build --release`, using:
```
debug = true
lto = "thin"
```
165.1 MB

## Stripping debug symbols (spoilers: this is the culprit for most of the size)
`cargo build --release`, using:
```
debug = true
lto = "thin"
strip = "symbols"
```
10.8 MB

## Turning "debug" off
`cargo build --release`, using:
```
debug = false
lto = "thin"
strip = "symbols"
```
10.79 MB

(the fact that this is so similar to the `strip = "symbols"` variant tells me that *most* of the cost/benefit of using debug is likely related to those symbols)

## Panic = "abort"
`cargo build --release`, using:
```
debug = false
lto = "thin"
strip = "symbols"
panic = "abort"
```
9.2 MB

# TL;DR

I hadn't submitted this as a PR because I figured the inclusion of `debug = true` in the release profile was intentional, but figured I'd share these experiments - modifying the build flags for the release profile could make a very significant different in binary size that clients of the API would need to take on.

**Comment by karencfv** [comment]
Hey @smklein this is really cool! Thanks a bunch for taking the time to experiment with different configuration options and writing up this issue.

I'll find out the reason behind `debug = true` and make changes if possible :)

**Comment by rmustacc** [comment]
What's the distinction between debuginfo and symbols? Losing the symbol table is going to make doing anything with panic=abort a huge pain. I expect most of the winnings with `strip=symbols` to come from `strip=debuginfo`.

**Comment by smklein** [comment]
From the [docs](https://doc.rust-lang.org/rustc/codegen-options/index.html):

![image](https://user-images.githubusercontent.com/3258857/170712144-ef790f7f-5bef-4687-aeb2-7487bee6028b.png)
![image](https://user-images.githubusercontent.com/3258857/170712229-1e78bb3a-ffbe-4e4d-b726-eb1d4bb50bad.png)

I'm not entirely sure what is meant by "the rest of the symbol table section", but I suspect that they're fairly similar based on the rest of the description.

**Comment by rmustacc** [comment]
I should have been clearer, I understand the difference from a rust perspective, I was asking abou the difference from a size perspective.

The actual distinction I can address. The symbol table usually refers to the standardish ELF `.symtab` section and its corresponding string table. Technically this is probably being determined through the use of looking at the ELF `sh_type` as equal to `SHT_SYMTAB` and the corresponding string table is in the section header's `sh_link`. The purpose of the symbol table is to give the name and length of each symbol in the program, whether public or private. This is how you can get things like `nm` to work and is required for resolving symbols to names. Without a symbol table we'd have to have someway of mapping the built artifact to something we distributed, which especially for the client CLI we shouldn't assume will happen as customers may prefer to build themselves. Because the stuff here was all looking at enabling `painc="abort"`, keeping the symbol table (which is usually quite small relative to the debug information) is pretty useful.

As an example, in illumos we never strip the symbol table out of binaries or libraries even though we don't ship debug information. This hasn't been a negative size impact for distributing things (though they were all built in C and we're not linking in libstd because of dynamic linking of libc). Including all the DWARF (e.g. on *nix the `debuginfo` bits) does have a large impact on binary size. In general, stripping the symbol table is pretty bad and is one of the reasons that debugging core dumps in most linux distros without special debug packages or controlling all the tooling and builds is a pain in the ass. Where as with a symbol table generally, if it ends up in the core dump (which at least it does on helios), then we can debug from just a core file.

**Comment by karencfv** [comment]
Thanks both! Looks like the binary sizes changed since this issue was opened. Building from the latest commit on the `main` branch the sizes that I'm getting are the following:

#### Debug build
`cargo build`

oxide binary is 62.8 MB

#### Release build
`cargo build --release`

oxide binary is 15 MB

#### Release build stripping debuginfo
`cargo build --release` using:

```
[profile.release]
debug = true
lto = "thin"
strip= "debuginfo"
```

oxide binary is 11.5 MB

It seems that with the latest changes, the difference between stripping debug info and keeping it is not as big. I realise we are trying to take up as little space as possible with our binaries. With these new sizes what are your opinions on keeping all the debug info vs not?

**Comment by karencfv** [comment]
Closing this as we are moving onto https://github.com/oxidecomputer/oxide-sdk-and-cli
